aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,575 @@
+                &mut secret_key_size,
+            )
+        } {
+            return Err(Unspecified);
+        }
+        Ok(PrivateKey {
+            algorithm: alg,
+            evp_pkey: kyber_key,
+            priv_key: priv_key_bytes.into(),
+        })
+    }
+
+    /// Return the algorithm associated with the given KEM private key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm {
+        self.algorithm
+    }
+
+    /// Computes the KEM public key from the KEM private key
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    pub fn compute_public_key(&self) -> Result<PublicKey, Unspecified> {
+        let mut public_key_size = self.algorithm.public_key_size();
+        let mut pubkey_bytes = vec![0u8; public_key_size];
+        if 1 != unsafe {
+            EVP_PKEY_get_raw_public_key(
+                self.evp_pkey.as_const_ptr(),
+                pubkey_bytes.as_mut_ptr(),
+                &mut public_key_size,
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        let pubkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_public_key(
+                self.algorithm.id.nid(),
+                pubkey_bytes.as_ptr(),
+                public_key_size,
+            )
+        })?;
+
+        Ok(PublicKey {
+            algorithm: self.algorithm,
+            evp_pkey: pubkey,
+            pub_key: pubkey_bytes.into(),
+        })
+    }
+
+    /// Performs the decapsulate operation using the current KEM private key on the given ciphertext.
+    ///
+    /// `ciphertext` is the ciphertext generated by the encapsulate operation using the KEM public key
+    /// associated with the current KEM private key.
+    ///
+    /// After the decapsulation is finished, `decapsulate` calls `kdf` with the raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error_value` when operation fails due to internal error.
+    #[allow(clippy::missing_panics_doc)]
+    pub fn decapsulate<F, R, E>(
+        &self,
+        ciphertext: &mut [u8],
+        error_value: E,
+        kdf: F,
+    ) -> Result<R, E>
+    where
+        F: FnOnce(&[u8]) -> Result<R, E>,
+    {
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) });
+        let ctx = if let Ok(ctx) = ctx {
+            ctx
+        } else {
+            return Err(error_value);
+        };
+
+        if 1 != unsafe {
+            EVP_PKEY_decapsulate(
+                *ctx,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+                ciphertext.as_mut_ptr(),
+                ciphertext.len(),
+            )
+        } {
+            return Err(error_value);
+        }
+
+        kdf(&shared_secret)
```

Comment:
where does this come from??

Reply:
`kdf` here is a user provided function of type `F` where is a `FnOnce(&[u8]) -> Result<R, E>`. The decapsulate flow the function is provided the plaintext of the secret, and the function can either return the secret as-is or do subsequent key-derivation using some other KDF of their choosing. The return signature of the provide function `F` is `Result<R, E>`, where `R` and `E` are generic parameters determined by the function provided by the user. This is the same result type that `decapsulate` returns. So to summarize: `decapsulate` takes a function provided by the user that is given the plaintext secret, and that function then influence the final result outcome of `decapsulate`, by either returning the secret as-is, performing further derivation logic, or returning an error.

This is similar for `encapsulate`, except that encapsulate takes a user function that excepts both the plaintext and ciphertext output from encaps algorithm, and that function can then determine what the final outputs are from the overall function.